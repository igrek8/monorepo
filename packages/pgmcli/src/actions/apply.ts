import { readFile } from 'fs/promises';
import { resolve } from 'path';
import pg from 'pg';

import { checkIntegrity } from '../core/check-integrity.js';
import { Config } from '../core/config.interface.js';
import { DefaultCommandOptions } from '../core/default-command-options.interface.js';
import { getAppliedMigrations } from '../core/get-applied-migrations.js';
import { getMigrations } from '../core/get-migrations.js';
import { LogLevel, getConsoleLevel, toServerSeverity } from '../core/logging.js';

export interface ApplyOptions extends DefaultCommandOptions {
  plan?: boolean;
  until?: string;
  meta?: string;
  tag: string;
  logLevel: LogLevel;
}

export async function apply(options: ApplyOptions, config?: Config, console = global.console) {
  const db = new pg.Client({
    ...config?.client,
    host: options.host,
    port: options.port,
    user: options.user,
    password: options.password,
    database: options.db,
  });
  db.on('notice', ({ severity, message }) => {
    const level = getConsoleLevel(severity);
    console[level](message);
  });
  const table = db.escapeIdentifier(options.table);
  const severity = toServerSeverity(options.logLevel);
  try {
    await db.connect();
    await db.query('BEGIN');
    await db.query(`SET client_min_messages TO ${severity}`);
    await db.query(`LOCK TABLE ${table} IN ACCESS EXCLUSIVE MODE`);
    const [migrations, applied] = await Promise.all([
      getMigrations(resolve(options.dir)),
      getAppliedMigrations(db, table),
    ]);
    checkIntegrity({ applied, migrations });
    if (options.until) {
      if (!migrations.has(options.until)) {
        throw new Error(`Migration ${options.until} not found`);
      }
      if (applied.has(options.until)) {
        return;
      }
    }
    for (const migration of migrations.values()) {
      if (applied.has(migration.id)) {
        continue;
      }
      console.info(`apply: ${migration.id}`);
      if (!options.plan) {
        const filePath = resolve(options.dir, migration.id);
        if (migration.id.endsWith('.sql')) {
          const content = await readFile(filePath, { encoding: 'utf-8' });
          const [up] = content.split(options.tag);
          if (up) {
            await db.query(up);
          }
        } else {
          const mod = await import(filePath);
          await (mod.up ?? mod.default.up)?.(db, { logLevel: options.logLevel });
        }
        const row = [migration.id, options.meta];
        await db.query(`INSERT INTO ${table} VALUES ($1, $2)`, row);
      }
      if (migration.id === options.until) {
        break;
      }
    }
    await db.query('COMMIT');
  } catch (e) {
    await db.query('ROLLBACK');
    throw e;
  } finally {
    await db.end();
  }
}
