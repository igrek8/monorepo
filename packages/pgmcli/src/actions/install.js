import { mkdir } from 'fs/promises';
import pg from 'pg';
const sql = `CREATE TABLE <table> (
  id VARCHAR(140) PRIMARY KEY,
  meta JSONB NULL CHECK (jsonb_typeof(meta) = 'object'),
  ts TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP
);
`;
export async function install(options, config) {
    const client = new pg.Client({
        ...config?.client,
        host: options.host,
        port: options.port,
        user: options.user,
        password: options.password,
        database: options.db,
    });
    const table = client.escapeIdentifier(options.table);
    try {
        await client.connect();
        await client.query(sql.replace('<table>', table));
        await mkdir(options.dir, { recursive: true });
    }
    finally {
        await client.end();
    }
}
//# sourceMappingURL=install.js.map