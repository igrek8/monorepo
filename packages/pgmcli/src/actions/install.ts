import { mkdir } from 'fs/promises';
import pg from 'pg';

import type { Config } from '../core/config.interface.js';
import type { DefaultCommandOptions } from '../core/default-command-options.interface.js';

export type InstallOptions = DefaultCommandOptions;

const sql = `CREATE TABLE <table> (
  id VARCHAR(140) PRIMARY KEY,
  meta JSONB NULL CHECK (jsonb_typeof(meta) = 'object'),
  ts TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP
);
`;

export async function install(options: InstallOptions, config?: Config) {
  const client = new pg.Client({
    ...config?.client,
    host: options.host,
    port: options.port,
    user: options.user,
    password: options.password,
    database: options.db,
  });
  const table = client.escapeIdentifier(options.table);
  try {
    await client.connect();
    await client.query(sql.replace('<table>', table));
    await mkdir(options.dir, { recursive: true });
  } finally {
    await client.end();
  }
}
