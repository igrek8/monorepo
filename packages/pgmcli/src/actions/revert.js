import { readFile } from 'fs/promises';
import { resolve } from 'path';
import pg from 'pg';
import { checkIntegrity } from '../core/check-integrity.js';
import { getAppliedMigrations } from '../core/get-applied-migrations.js';
import { getMigrations } from '../core/get-migrations.js';
import { getConsoleLevel, toServerSeverity } from '../core/logging.js';
export async function revert(options, config, console = global.console) {
    const db = new pg.Client({
        ...config?.client,
        host: options.host,
        port: options.port,
        user: options.user,
        password: options.password,
        database: options.db,
    });
    db.on('notice', ({ severity, message }) => {
        const level = getConsoleLevel(severity);
        console[level](message);
    });
    const table = db.escapeIdentifier(options.table);
    const severity = toServerSeverity(options.logLevel);
    try {
        await db.connect();
        await db.query('BEGIN');
        await db.query(`SET client_min_messages TO ${severity}`);
        await db.query(`LOCK TABLE ${table} IN ACCESS EXCLUSIVE MODE`);
        const [migrations, applied] = await Promise.all([
            getMigrations(resolve(options.dir)),
            getAppliedMigrations(db, table),
        ]);
        checkIntegrity({ applied, migrations });
        if (!migrations.has(options.until)) {
            throw new Error(`Migration ${options.until} not found`);
        }
        if (!applied.has(options.until)) {
            return;
        }
        for (const migration of Array.from(migrations.values()).reverse()) {
            if (!applied.has(migration.id)) {
                continue;
            }
            console.info(`revert: ${migration.id}`);
            if (!options.plan) {
                const filePath = resolve(options.dir, migration.id);
                if (migration.id.endsWith('.sql')) {
                    const content = await readFile(filePath, { encoding: 'utf-8' });
                    const [, down] = content.split(options.tag);
                    if (down) {
                        await db.query(down);
                    }
                }
                else {
                    const module = await import(filePath);
                    await (module.down ?? module.default.down)?.(db, { logLevel: options.logLevel });
                }
                await db.query(`DELETE FROM ${table} WHERE id = $1`, [migration.id]);
            }
            if (options.until === migration.id) {
                break;
            }
        }
        await db.query('COMMIT');
    }
    catch (e) {
        await db.query('ROLLBACK');
        throw e;
    }
    finally {
        await db.end();
    }
}
//# sourceMappingURL=revert.js.map