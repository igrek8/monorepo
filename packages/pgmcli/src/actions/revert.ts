import { readFileSync } from 'fs';
import { resolve } from 'path';
import { Client } from 'pg';
import { checkIntegrity } from '../core/checkIntegrity';
import type { Config } from '../core/Config';
import type { DefaultCommandOptions } from '../core/DefaultCommandOptions';
import { getAppliedMigrations } from '../core/getAppliedMigrations';
import { getMigrations } from '../core/getMigrations';
import { LogLevel, getConsoleLevel, toServerSeverity } from '../core/logging';

export interface RevertOptions extends DefaultCommandOptions {
  plan?: boolean;
  until: string;
  meta?: string;
  tag: string;
  logLevel: LogLevel;
}

export async function revert(options: RevertOptions, config?: Config, console = globalThis.console) {
  const db = new Client({
    ...config?.client,
    host: options.host,
    port: options.port,
    user: options.user,
    password: options.password,
    database: options.db,
  });
  db.on('notice', ({ severity, message }) => {
    const level = getConsoleLevel(severity);
    console[level](message);
  });
  const table = db.escapeIdentifier(options.table);
  const severity = toServerSeverity(options.logLevel);
  try {
    await db.connect();
    await db.query('BEGIN');
    await db.query(`SET client_min_messages TO ${severity}`);
    await db.query(`LOCK TABLE ${table} IN ACCESS EXCLUSIVE MODE`);
    const migrations = getMigrations(resolve(options.dir));
    const applied = await getAppliedMigrations(db, table);
    checkIntegrity(migrations, applied);
    if (!migrations.has(options.until)) {
      throw new Error(`Migration ${options.until} not found`);
    }
    if (!applied.has(options.until)) {
      return;
    }
    for (const migration of Array.from(migrations.values()).reverse()) {
      if (!applied.has(migration.id)) {
        continue;
      }
      console.info(`revert: ${migration.id}`);
      if (!options.plan) {
        const filePath = resolve(options.dir, migration.id);
        if (migration.id.endsWith('.sql')) {
          const content = readFileSync(filePath, { encoding: 'utf-8' });
          const [, down] = content.split(options.tag);
          if (down) {
            await db.query(down);
          }
        } else {
          const module = await import(filePath);
          await (module.down ?? module.default.down)?.(db, { logLevel: options.logLevel });
        }
        await db.query(`DELETE FROM ${table} WHERE id = $1`, [migration.id]);
      }
      if (options.until === migration.id) {
        break;
      }
    }
    await db.query('COMMIT');
  } catch (e) {
    await db.query('ROLLBACK');
    throw e;
  } finally {
    await db.end();
  }
}
