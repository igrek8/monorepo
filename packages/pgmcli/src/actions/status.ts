import { resolve } from 'path';
import pg from 'pg';

import type { Config } from '../core/config.interface.js';
import type { DefaultCommandOptions } from '../core/default-command-options.interface.js';
import { getAppliedMigrations } from '../core/get-applied-migrations.js';
import { getMigrations } from '../core/get-migrations.js';

export type StatusOptions = DefaultCommandOptions;

export async function status(options: StatusOptions, config?: Config, console = globalThis.console) {
  const client = new pg.Client({
    ...config?.client,
    host: options.host,
    port: options.port,
    user: options.user,
    password: options.password,
    database: options.db,
  });
  const table = client.escapeIdentifier(options.table);
  try {
    await client.connect();
    await client.query('BEGIN');
    await client.query(`LOCK TABLE ${table} IN ACCESS EXCLUSIVE MODE`);
    const [migrations, applied] = await Promise.all([
      getMigrations(resolve(options.dir)),
      getAppliedMigrations(client, table),
    ]);
    migrations.forEach(({ id }) => {
      const status = applied.has(id) ? 'applied' : 'pending';
      const meta = applied.get(id)?.meta;
      let message = `${status}: ${id}`;
      if (meta) message += ` ${JSON.stringify(meta, null, 2)}`;
      console.info(message);
    });
  } finally {
    await client.query('ROLLBACK');
    await client.end();
  }
}
