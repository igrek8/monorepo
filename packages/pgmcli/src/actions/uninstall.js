import pg from 'pg';
export async function uninstall(options, config) {
    const client = new pg.Client({
        ...config?.client,
        host: options.host,
        port: options.port,
        user: options.user,
        password: options.password,
        database: options.db,
    });
    try {
        await client.connect();
        const table = client.escapeIdentifier(options.table);
        await client.query(`DROP TABLE ${table}`);
    }
    finally {
        await client.end();
    }
}
//# sourceMappingURL=uninstall.js.map