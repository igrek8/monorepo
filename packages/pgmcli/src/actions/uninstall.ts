import pg from 'pg';

import type { Config } from '../core/config.interface.js';
import { DefaultCommandOptions } from '../core/default-command-options.interface.js';

export type UninstallOptions = DefaultCommandOptions;

export async function uninstall(options: UninstallOptions, config?: Config) {
  const client = new pg.Client({
    ...config?.client,
    host: options.host,
    port: options.port,
    user: options.user,
    password: options.password,
    database: options.db,
  });
  try {
    await client.connect();
    const table = client.escapeIdentifier(options.table);
    await client.query(`DROP TABLE ${table}`);
  } finally {
    await client.end();
  }
}
