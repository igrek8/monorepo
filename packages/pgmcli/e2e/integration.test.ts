import type { PathLike } from 'fs';
import fs from 'fs/promises';
import { tmpdir } from 'os';
import path from 'path';
import { Client } from 'pg';
import { PassThrough } from 'stream';
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest';
import { apply } from '../src/actions/apply';
import { create } from '../src/actions/create';
import { install } from '../src/actions/install';
import { revert } from '../src/actions/revert';
import { status } from '../src/actions/status';
import { uninstall } from '../src/actions/uninstall';
import type { Config } from '../src/core/config.interface';
import { DEFAULT_HOST, DEFAULT_PORT, DEFAULT_TAG } from '../src/core/constants';
import type { DefaultCommandOptions } from '../src/core/default-command-options.interface';
import { LogLevel } from '../src/core/logging';

async function getFileContent(path: PathLike) {
  return (await fs.readFile(path, 'utf-8')).trim();
}

const stdout = new PassThrough();
const stderr = new PassThrough();
const logger = new console.Console(stdout, stderr);
const info = vi.spyOn(logger, 'info');

const options: DefaultCommandOptions = {
  host: process.env.POSTGRES_HOST ?? DEFAULT_HOST,
  port: Number(process.env.POSTGRES_PORT ?? DEFAULT_PORT),
  user: 'postgres',
  table: 'migrations',
  dir: path.join(__dirname, '..', 'migrations'),
};

const config: Config = {
  client: {
    ssl: false,
  },
};

const tables = ['payment', 'order_item', 'order', 'book', 'customer', 'migrations'];

async function listTables(db: Client): Promise<string[]> {
  const { rows } = await db.query(
    `
    SELECT table_name
    FROM information_schema.tables
    WHERE table_schema = $1
    ORDER BY table_name ASC
  `,
    ['public'],
  );
  return rows.map((row) => row.table_name);
}

describe.sequential('pgmcli', () => {
  let db: Client;

  beforeAll(async () => {
    db = new Client({
      host: options.host,
      port: options.port,
      user: options.user,
      database: options.db,
    });
    await db.connect();
    await Promise.all(tables.map((table) => db.query(`DROP TABLE IF EXISTS "${table}"`)));
  });

  afterAll(async () => {
    await db.end();
  });

  describe('install', () => {
    it('creates migrations table', async () => {
      await install(options);
      await expect(listTables(db)).resolves.toStrictEqual(['migrations']);
    });
  });

  describe('status', () => {
    it('shows the status of the migrations', async () => {
      await status(options, config, logger);
      expect(info).toHaveBeenNthCalledWith(1, 'pending: 1_add_customer_table.sql');
      expect(info).toHaveBeenNthCalledWith(2, 'pending: 2_add_book_table.sql');
      expect(info).toHaveBeenNthCalledWith(3, 'pending: 3_add_order_table.js');
      expect(info).toHaveBeenNthCalledWith(4, 'pending: 4_add_order_item_table.sql');
      expect(info).toHaveBeenNthCalledWith(5, 'pending: 5_add_payment_table.sql');
    });
  });

  describe('apply', () => {
    it('applies the first 3 migrations', async () => {
      info.mockReset();
      await apply(
        {
          ...options,
          until: '3_add_order_table.js',
          tag: DEFAULT_TAG,
          logLevel: LogLevel.INFO,
        },
        config,
        logger,
      );
      await expect(listTables(db)).resolves.toStrictEqual(['book', 'customer', 'migrations', 'order']);
      expect(info).toHaveBeenCalledWith('Creating table book');
    });

    it('skips if until migration is already applied', async () => {
      info.mockReset();
      await apply(
        {
          ...options,
          tag: DEFAULT_TAG,
          until: '3_add_order_table.js',
          logLevel: LogLevel.INFO,
        },
        config,
        logger,
      );
      await expect(listTables(db)).resolves.toStrictEqual(['book', 'customer', 'migrations', 'order']);
      expect(info).not.toHaveBeenCalled();
    });

    it('throws if until migration is not found', async () => {
      await expect(
        apply(
          {
            ...options,
            until: '6_add_payment_term_table.sql',
            tag: DEFAULT_TAG,
            logLevel: LogLevel.INFO,
          },
          config,
          logger,
        ),
      ).rejects.toThrow('Migration 6_add_payment_term_table.sql not found');
    });

    it('shows the status of the migrations', async () => {
      info.mockReset();
      await status(options, config, logger);
      expect(info).toHaveBeenNthCalledWith(1, 'applied: 1_add_customer_table.sql');
      expect(info).toHaveBeenNthCalledWith(2, 'applied: 2_add_book_table.sql');
      expect(info).toHaveBeenNthCalledWith(3, 'applied: 3_add_order_table.js');
      expect(info).toHaveBeenNthCalledWith(4, 'pending: 4_add_order_item_table.sql');
      expect(info).toHaveBeenNthCalledWith(5, 'pending: 5_add_payment_table.sql');
    });

    it('applies the next 2 migrations', async () => {
      await apply(
        {
          ...options,
          tag: DEFAULT_TAG,
          logLevel: LogLevel.INFO,
          meta: '{ "test": true }',
        },
        config,
        logger,
      );
      await expect(listTables(db)).resolves.toStrictEqual([
        'book',
        'customer',
        'migrations',
        'order',
        'order_item',
        'payment',
      ]);
    });

    it('shows the status of the migrations', async () => {
      info.mockReset();
      await status(options, config, logger);
      expect(info).toHaveBeenNthCalledWith(1, 'applied: 1_add_customer_table.sql');
      expect(info).toHaveBeenNthCalledWith(2, 'applied: 2_add_book_table.sql');
      expect(info).toHaveBeenNthCalledWith(3, 'applied: 3_add_order_table.js');
      expect(info).toHaveBeenNthCalledWith(4, 'applied: 4_add_order_item_table.sql {\n  "test": true\n}');
      expect(info).toHaveBeenNthCalledWith(5, 'applied: 5_add_payment_table.sql {\n  "test": true\n}');
    });
  });

  describe('revert', () => {
    it('reverts the last 2 migrations', async () => {
      await revert(
        {
          ...options,
          until: '4_add_order_item_table.sql',
          tag: DEFAULT_TAG,
          logLevel: LogLevel.INFO,
        },
        config,
        logger,
      );
      await expect(listTables(db)).resolves.toStrictEqual(['book', 'customer', 'migrations', 'order']);
    });

    it('skips if until migration is already reverted', async () => {
      info.mockReset();
      await revert(
        {
          ...options,
          until: '4_add_order_item_table.sql',
          tag: DEFAULT_TAG,
          logLevel: LogLevel.INFO,
        },
        config,
        logger,
      );
      await expect(listTables(db)).resolves.toStrictEqual(['book', 'customer', 'migrations', 'order']);
      expect(info).not.toHaveBeenCalled();
    });

    it('throws if until migration is not found', async () => {
      await expect(
        revert(
          {
            ...options,
            until: '6_add_payment_term_table.sql',
            tag: DEFAULT_TAG,
            logLevel: LogLevel.INFO,
          },
          config,
          logger,
        ),
      ).rejects.toThrow('Migration 6_add_payment_term_table.sql not found');
    });

    it('shows the status of the migrations', async () => {
      info.mockReset();
      await status(options, config, logger);
      expect(info).toHaveBeenNthCalledWith(1, 'applied: 1_add_customer_table.sql');
      expect(info).toHaveBeenNthCalledWith(2, 'applied: 2_add_book_table.sql');
      expect(info).toHaveBeenNthCalledWith(3, 'applied: 3_add_order_table.js');
      expect(info).toHaveBeenNthCalledWith(4, 'pending: 4_add_order_item_table.sql');
      expect(info).toHaveBeenNthCalledWith(5, 'pending: 5_add_payment_table.sql');
    });

    it('reverts the last 3 migrations', async () => {
      info.mockReset();
      await revert(
        {
          ...options,
          until: '1_add_customer_table.sql',
          tag: DEFAULT_TAG,
          logLevel: LogLevel.INFO,
        },
        config,
        logger,
      );
      await expect(listTables(db)).resolves.toStrictEqual(['migrations']);
      expect(info).toHaveBeenCalledWith('Dropping table book');
    });

    it('shows the status of the migrations', async () => {
      info.mockReset();
      await status(options, config, logger);
      expect(info).toHaveBeenNthCalledWith(1, 'pending: 1_add_customer_table.sql');
      expect(info).toHaveBeenNthCalledWith(2, 'pending: 2_add_book_table.sql');
      expect(info).toHaveBeenNthCalledWith(3, 'pending: 3_add_order_table.js');
      expect(info).toHaveBeenNthCalledWith(4, 'pending: 4_add_order_item_table.sql');
      expect(info).toHaveBeenNthCalledWith(5, 'pending: 5_add_payment_table.sql');
    });
  });

  describe('uninstall', () => {
    it('removes the migrations table', async () => {
      await uninstall(options, config);
      await expect(listTables(db)).resolves.toStrictEqual([]);
    });
  });

  describe('create', () => {
    it('creates a new migration file', async () => {
      const outDir = tmpdir();
      const outFile = path.join(outDir, '1704067200000_add_new_table.sql');
      vi.setSystemTime('2024-01-01T00:00:00Z');
      await create({ dir: outDir, name: 'add_new_table.sql', tag: DEFAULT_TAG }, logger);
      await expect(getFileContent(outFile)).resolves.toMatchSnapshot();
    });
  });
});
