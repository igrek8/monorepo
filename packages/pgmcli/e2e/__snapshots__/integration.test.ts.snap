// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pgmcli > create > creates a new migration file 1`] = `
"DO $$
BEGIN
  RAISE LOG 'Applying';
END $$;

-- DO NOT REMOVE - THIS LINE SEPARATES APPLY AND REVERT OPERATIONS. REVERT BEGIN

DO $$
BEGIN
  RAISE LOG 'Reverting';
END $$;"
`;
